File: backend/app/schemas/ml_lab.py
"""Schemi Pydantic per la pagina tecnica ML Lab (training / test / eval)."""

from __future__ import annotations

from pydantic import BaseModel, Field


class ModelInfo(BaseModel):
    """Stato di un modello addestrato da noi."""

    key: str                      # "condition-mlp" | "gap-mlp"
    name: str
    nature: str                   # "own" — per il codice colore in UI
    task: str                     # descrizione breve del compito
    available: bool               # pesi/adapter presenti su disco
    architecture: str | None = None
    metrics: dict | None = None   # metriche salvate nel checkpoint al training
    labels: list[str] | None = None


class DatasetInfo(BaseModel):
    key: str                      # "garment_condition" | "wardrobe"
    name: str
    available: bool
    n_samples: int | None = None
    detail: str | None = None


class MlLabStatus(BaseModel):
    models: list[ModelInfo]
    datasets: list[DatasetInfo]


class ConditionPredictOut(BaseModel):
    condition: str
    confidence: float
    probabilities: dict[str, float]


class GapPredictIn(BaseModel):
    """Conteggi per categoria + contesto colore/uso, per il simulatore what-if."""

    counts: dict[str, int] = Field(default_factory=dict)
    n_colors: int = Field(3, ge=0, le=20)
    has_neutral: bool = True
    ghost_ratio: float = Field(0.2, ge=0.0, le=1.0)


class GapPredictOut(BaseModel):
    gaps: list[str]                    # codici label attive
    labels: dict[str, str]             # code -> etichetta leggibile
    probabilities: dict[str, float]    # vuote se fallback a regole
    balanced: bool
    source: str                        # "neural-net" | "rules"

File: frontend/src/api/mlLab.ts
/** Client per la pagina tecnica ML Lab (stato reti, prova interattiva). */
import { API_BASE, asError, jsonOrThrow } from './client'

export interface ModelInfo {
  key: string
  name: string
  /** "own" (rete nostra) | "gen" (generativa) — codice colore in UI */
  nature: string
  task: string
  available: boolean
  architecture: string | null
  metrics: Record<string, unknown> | null
  labels: string[] | null
}

export interface DatasetInfo {
  key: string
  name: string
  available: boolean
  n_samples: number | null
  detail: string | null
}

export interface MlLabStatus {
  models: ModelInfo[]
  datasets: DatasetInfo[]
}

export interface ConditionPredictOut {
  condition: string
  confidence: number
  probabilities: Record<string, number>
}

export interface GapPredictIn {
  counts: Record<string, number>
  n_colors: number
  has_neutral: boolean
  ghost_ratio: number
}

export interface GapPredictOut {
  gaps: string[]
  labels: Record<string, string>
  probabilities: Record<string, number>
  balanced: boolean
  source: string
}

export async function getMlLabStatus(): Promise<MlLabStatus> {
  return jsonOrThrow<MlLabStatus>(await fetch(`${API_BASE}/ml/models`))
}

export async function predictCondition(image: File): Promise<ConditionPredictOut> {
  const fd = new FormData()
  fd.append('image', image)
  const r = await fetch(`${API_BASE}/ml/condition/predict`, { method: 'POST', body: fd })
  if (!r.ok) throw await asError(r)
  return (await r.json()) as ConditionPredictOut
}

export async function predictGap(input: GapPredictIn): Promise<GapPredictOut> {
  return jsonOrThrow<GapPredictOut>(
    await fetch(`${API_BASE}/ml/gap/predict`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(input),
    }),
  )
}

export function confusionMatrixUrl(): string {
  return `${API_BASE}/ml/condition/confusion-matrix`
}
